tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +224,286 @@ static bool HasCanonicalDimensionNumbers(
   return true;
 }
 
+//===----------------------------------------------------------------------===//
+// mhlo.Einsum conversion patterns.
+//===----------------------------------------------------------------------===//
+
+// Looks through a set of dimension that has been marked as Reduction axes
+// if it is found within the set, then we set it as "reduction", otherwise
+// we can label it as "parallel"
+SmallVector<StringRef, 3> GetEinsumLoopsAttrs(
+    const llvm::SmallSetVector<StringRef, 4>& input_ind,
+    const llvm::SmallSetVector<StringRef, 4>& reduction_dims) {
+  SmallVector<StringRef, 3> res;
+  for (const StringRef& dim : input_ind) {
+    if (!reduction_dims.contains(dim)) {
+      res.push_back(getParallelIteratorTypeName());
+    } else {
+      res.push_back(getReductionIteratorTypeName());
+    }
+  }
+  return res;
+}
+
+// Check for where each dimension of the output operand came from
+// If that dimension is dynamically sized then we push it to dyn_size vector
+SmallVector<Value, 2> ExtractDynamicEinsumSizes(
+    OpBuilder& b, const Location& loc, const Value& lhs, const Value& rhs,
```

Comment:
We don't need const reference for `Location` and `Value`s.
```suggestion
    OpBuilder& b, Location loc, Value lhs, Value rhs,
```

Reply:
done